test_communication/test/test_messages_c.cpp

```diff
@@ -0,0 +1,1127 @@
+  template<> \
+  void fini_message(TYPE * msg) { \
+    TYPE ## __fini(msg); \
+  }
+
+// Test publish/subscribe with a variety of messages
+
+// Define functions and test cases for each message type
+template<>
+size_t get_message_num(test_communication__msg__Primitives * msg)
+{
+  (void)msg;
+  return 4;
+}
+
+template<>
+void init_message(test_communication__msg__Primitives * msg)
+{
+  test_communication__msg__Primitives__init(msg);
+}
+
+template<>
+void get_message(test_communication__msg__Primitives * msg, size_t msg_num)
+{
+  test_communication__msg__Primitives__init(msg);
+  switch (msg_num) {
+    case 0:
+      msg->bool_value = false;
+      msg->byte_value = 0;
+      msg->char_value = '\0';
+      msg->float32_value = 0.0f;
+      msg->float64_value = 0;
+      msg->int8_value = 0;
+      msg->uint8_value = 0;
+      msg->int16_value = 0;
+      msg->uint16_value = 0;
+      msg->int32_value = 0;
+      msg->uint32_value = 0;
+      msg->int64_value = 0;
+      msg->uint64_value = 0;
+      rosidl_generator_c__String__assign(&msg->string_value, "");
+      break;
+    case 1:
+      msg->bool_value = true;
+      msg->byte_value = 255;
+      msg->char_value = '\x7f';
+      msg->float32_value = 1.125f;
+      msg->float64_value = 1.125;
+      msg->int8_value = (std::numeric_limits<int8_t>::max)();
+      msg->uint8_value = (std::numeric_limits<uint8_t>::max)();
+      msg->int16_value = (std::numeric_limits<int16_t>::max)();
+      msg->uint16_value = (std::numeric_limits<uint16_t>::max)();
+      msg->int32_value = (std::numeric_limits<int32_t>::max)();
+      msg->uint32_value = (std::numeric_limits<uint32_t>::max)();
+      msg->int64_value = (std::numeric_limits<int64_t>::max)();
+      msg->uint64_value = (std::numeric_limits<uint64_t>::max)();
+      rosidl_generator_c__String__assign(&msg->string_value, "max value");
+      break;
+    case 2:
+      msg->bool_value = false;
+      msg->byte_value = 0;
+      msg->char_value = 0x0;
+      msg->float32_value = -2.125f;
+      msg->float64_value = -2.125;
+      msg->int8_value = (std::numeric_limits<int8_t>::min)();
+      msg->uint8_value = 0;
+      msg->int16_value = (std::numeric_limits<int16_t>::min)();
+      msg->uint16_value = 0;
+      msg->int32_value = (std::numeric_limits<int32_t>::min)();
+      msg->uint32_value = 0;
+      msg->int64_value = (std::numeric_limits<int64_t>::min)();
+      msg->uint64_value = 0;
+      rosidl_generator_c__String__assign(&msg->string_value, "min value");
+      break;
+    case 3:
+      msg->bool_value = true;
+      msg->byte_value = 1;
+      msg->char_value = '\1';
+      msg->float32_value = 1.0f;
+      msg->float64_value = 1;
+      msg->int8_value = 1;
+      msg->uint8_value = 1;
+      msg->int16_value = 1;
+      msg->uint16_value = 1;
+      msg->int32_value = 1;
+      msg->uint32_value = 1;
+      msg->int64_value = 1;
+      msg->uint64_value = 1;
+      char string_value[200] = {};
```

Comment:
The tests in other languages explicitly try a larger size since <= 255 might work but not for higher numbers.